src/eins-network-id.c

```diff
@@ -0,0 +1,260 @@
+/* Copyright 2017 Endless Mobile, Inc.
+ *
+ * This file is part of eos-metrics-instrumentation.
+ *
+ * eos-metrics-instrumentation is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * eos-metrics-instrumentation is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-metrics-instrumentation.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <glib.h>
+#include <stdio.h>
+
+#include "eins-network-id.h"
+
+#define PROC_IPV4_ROUTE_FILE "/proc/net/route"
+#define IPV4_ROUTE_REGEXP "^(?:\\S+\\s+){2}(?P<Gateway>[0-9A-Fa-f]{8})\\s+(?:\\S+\\s+){4}(?P<Mask>00000000)"
+
+/*
+ * greps the /proc/net/route file for the IPv4 default route. note that even
+ * on machines with multiple interfaces, the first listed 00000000 route will
+ * be the first one selected for outbound traffic, so we don’t need to do any
+ * more complex searching/sorting here (verified in net/ipv4/fib_trie.c on
+ * kernel 4.10.5 - the trie is ordered by (prefix,tos,priority) so the highest
+ * preference default route will appear first)
+ */
+static gchar *
+get_ipv4_default_gateway (void)
+{
+  g_autofree gchar *route = NULL;
+  g_autoptr(GError) err = NULL;
+
+  if (!g_file_get_contents (PROC_IPV4_ROUTE_FILE, &route, NULL, &err))
+    {
+      g_debug ("could not load " PROC_IPV4_ROUTE_FILE ": %s", err->message);
+      return NULL;
+    }
+
+  g_autoptr(GRegex) route_rx = g_regex_new (IPV4_ROUTE_REGEXP, G_REGEX_MULTILINE, 0, NULL);
+  g_assert (route_rx != NULL);
+
+  g_autoptr(GMatchInfo) route_mi = NULL;
+  if (!g_regex_match (route_rx, route, 0, &route_mi))
+    {
+      g_debug ("couldn’t match IPv4 default gateway in " PROC_IPV4_ROUTE_FILE);
+      return NULL;
+    }
+
+  const gchar *gateway = g_match_info_fetch_named (route_mi, "Gateway");
+  g_assert (gateway != NULL);
+
+  /* the input to this g_ascii_strtoull call should already be
+   * rigorously validated by the regexp: failure is not an option */
+  gchar *endptr = NULL;
+  errno = 0;
+  guint32 gateway_ip = (guint32) g_ascii_strtoull (gateway, &endptr, 16);
+  g_assert (gateway + 8 == endptr);
+  g_assert (gateway_ip != 0 || errno == 0);
```

Comment:
Also, I think you should drop the `gateway_ip != 0` clause: you should only care about `errno` and `endptr` for the error reporting, since the return value can be `0` *or* `G_MAXUINT64` in different error cases.

i.e. `g_ascii_strtoull()` is a terribly designed function.